pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/channel/InMemoryChannel.java

```diff
@@ -0,0 +1,51 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.mailbox.channel;
+
+import java.util.concurrent.ArrayBlockingQueue;
+
+
+/**
+ * Used by {@link org.apache.pinot.query.mailbox.InMemoryMailboxService} for passing data between stages when sender
+ * and receiver are in the same process.
+ */
+public class InMemoryChannel<T> {
+  private final ArrayBlockingQueue<T> _channel;
+  private final String _hostname;
+  private final int _port;
+  private boolean _completed = false;
+
+  public InMemoryChannel(ArrayBlockingQueue<T> channel, String hostname, int port) {
+    _channel = channel;
+    _hostname = hostname;
+    _port = port;
+  }
+
+  public ArrayBlockingQueue<T> getChannel() {
+    return _channel;
+  }
```

Comment:
IMO technically speaking we don't really need to keep this as a rapper of the array blocking queue b/c it can be directly coded inside the InMemoryMailboxService. 
but i would let others also take a look and see if this abstract makes it easier to understand

Reply:
In `MailboxReceiveOperator` we check if the mailbox is closed. For `InMemoryReceivingMailbox`, we need to confirm that:

1. The sender is done sending. When the sender is done, it will call complete here.
2. The ArrayBlockingQueue is empty.

If both of these are true, then we say that the mailbox is closed. I think this is actually equivalent to `MailboxContentStreamObserver` and not the channel itself.

Reply:
so in GRPC mailbox the mailbox itself doesn't keep the info regarding a complete signal. which make sense b/c the sender is on RPC. 
- for example the last metadatablock might get lost during the transport 
- GRPC helps to make sure the `onComplete()` method is called.

for in-memory mailbox, one can simply check whether a metadatablock is received to indicate whether it is completed (as you shall never lose a block on the ArrayBlockingQueue)